joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/passes/FullNameUniquenessPass.scala

```diff
@@ -0,0 +1,134 @@
+import io.shiftleft.codepropertygraph.generated.nodes.Method
+import io.shiftleft.codepropertygraph.generated.nodes.NamespaceBlock
+import io.shiftleft.codepropertygraph.generated.nodes.TypeDecl
+import io.shiftleft.passes.CpgPass
+import io.shiftleft.semanticcpg.language.*
+import io.shiftleft.semanticcpg.language.types.structure.NamespaceTraversal
+import org.slf4j.LoggerFactory
+
+/** Ensures that method, type decl, and namespace block fullNames are unique across the CPG. In C and C ++ code, these
+  * elements can appear multiple times across different translation units, e.g., with static functions in different
+  * files sharing the same name, classes declared and defined across different header and implementation files, or
+  * namespaces spanning different files. Sometimes we also may fail to resolve proper type names (missing includes,
+  * missing macro expanding) and that, ultimately, leads to the same fullName being generated.
+  *
+  * The CPG specification requires unique fullNames for all kind of linking and querying operations.
+  *
+  * This unique fullName generation cannot be performed during the initial AST traversal because:
+  *   - The AST is built on a per-file basis where we don't have knowledge of methods defined in other files
+  *   - Name uniqueness must be established across the entire codebase, not just within a single file.
+  *   - Using a global cache during parsing would slow down our parallel processing.
+  *   - Simply prepending the filename to the fullName does not work as we can not determine that fullName for all calls
+  *     (leads to linking issues later on) or referenced type. Also, the same fullName may be generated multiple times
+  *     in on file due to missing type information (as mentioned above).
+  *
+  * Instead , this pass runs after all methods, type decls, and namespaces have been added to the CPG, identifies
+  * duplicates, and appends unique suffixes to ensure each one can be uniquely identified. Calls to static methods
+  * affected by this are also updated to reflect the new method fullName.
+  */
+class FullNameUniquenessPass(cpg: Cpg) extends CpgPass(cpg) {
+
+  private type AffectedNodeType = Method | TypeDecl | NamespaceBlock
+  private val logger = LoggerFactory.getLogger(classOf[FullNameUniquenessPass])
+
+  override def run(dstGraph: DiffGraphBuilder): Unit = {
+    handleMethods(dstGraph)
+    handleTypeDecls(dstGraph)
+    handleNamespaceBlocks(dstGraph)
+  }
+
+  private def handleMethods(dstGraph: DiffGraphBuilder): Unit = {
+    val methodFullNameMap = generateStableFullNameMapping(cpg.method.nameNot(NamespaceTraversal.globalNamespaceName))
+    methodFullNameMap.foreach { case (fullName, methods) =>
+      logDuplicates(methods, fullName)
+      val sortedMethods = sortNodesByLocation(methods)
+      lazy val callsAffected =
+        cpg.call.methodFullNameExact(fullName).filterNot(_.file.exists(_.name == sortedMethods.head.filename)).l
```

Comment:
I think `call.file` uses `CONTAINS` edges which won't be available at the time this is running.

Reply:
It uses astIn which is fine.

Reply:
Huh, that's weird. But indeed makes it fine for us here.